keep/step/step.py

```diff
@@ -1,51 +1,198 @@
+        # 2. no "if" block is supplied, then use the AND between all conditions
+        if self.config.get("if"):
+            if_conf = self.config.get("if")
+        else:
+            # create a string of all conditions, separated by "and"
+            if_conf = " and ".join(
+                [f"{{{{ {condition.condition_alias} }}}} " for condition in conditions]
+            )
+
+        # Now check it
+        if if_conf:
+            if_met = self.io_handler.render(if_conf)
+            # Evaluate the condition string
+            if_met = eval(if_met)
+        else:
+            if_met = True
+
+        if not if_met:
+            self.logger.info(
+                "Action %s evaluated NOT to run, Reason: %s evaluated to false.",
+                self.config.get("name"),
+                if_conf,
+            )
+            return
 
+        self.logger.info("Action %s evaluated to run!", self.config.get("name"))
 
+        # Third, check throttling
+        # Now check if throttling is enabled
+        throttled = self._check_throttling(self.config.get("name"))
+        if throttled:
+            self.logger.info("Action %s is throttled", self.config.get("name"))
+            return
+
+        # Last, run the action
+        rendered_value = self.io_handler.render_context(self.provider_context)
+        # if the provider is async, run it in a new event loop
+        if inspect.iscoroutinefunction(self.provider.notify):
+            result = self._run_single_async()
+        # else, just run the provider
+        else:
+            if self.step_type == StepType.STEP:
+                try:
+                    rendered_providers_parameters = {}
+                    for parameter in self.provider_parameters:
+                        rendered_providers_parameters[parameter] = self.io_handler.render(
+                            self.provider_parameters[parameter]
+                        )
+                    step_output =  self.provider.query(**rendered_value)
+                    extra_context = self.provider.expose()
+                    rendered_providers_parameters.update(extra_context)
+                    self.context_manager.set_step_provider_paremeters(
+                        self.step_id, rendered_providers_parameters
+                    )
+                except Exception as e:
+                    raise StepError(e)
+                
+                return step_output
+            self.provider.notify(**rendered_value)
+            return True
+        
+    def _run_single_async(self):
+        """For async providers, run them in a new event loop
+
+        Raises:
+            ActionError: _description_
+        """
+        rendered_value = self.io_handler.render_context(self.provider_context)
+        # This is "magically solved" because of nest_asyncio but probably isn't best practice
+        loop = asyncio.new_event_loop()
+        task = loop.create_task(self.provider.notify(**rendered_value))
```

Comment:
```suggestion
        if self.step_type == StepType.STEP:
            task = loop.create_task(self.provider.query(**rendered_value))
        else:
            task = loop.create_task(self.provider.notify(**rendered_value))
```

Reply:
Ok. I will add that. Thank you for pointing that out.